src/aspire/utils/rotation.py

```diff
@@ -0,0 +1,280 @@
+                num_rots == angles.shape[0],
+                "Number of rotation matrices should equal to "
+                "number of sets of Euler angles.",
+            )
+            self.angles = angles.astype(self.dtype)
+        else:
+            self.angles = self._uniform_random_angles(num_rots, seed=seed)
+        self.num_rots = num_rots
+
+        self.data = self.rot_matrices
+        self.shape = (num_rots, 3, 3)
+
+    def __str__(self):
+        """
+        String representation.
+        """
+        return (
+            f"Rotation object with matrices[{self.num_rots}, 3, 3] of {self.dtype} type"
+        )
+
+    def __len__(self):
+        return self.num_rots
+
+    def __getitem__(self, item):
+        return self.data[item]
+
+    def __setitem__(self, key, value):
+        self.data[key] = value
+
+    def __mul__(self, other):
+        if isinstance(other, Rotation):
+            output = self.data @ other.data
+        else:
+            output = self.data * other
+        return output
+
+    def _uniform_random_angles(self, n, seed=None):
+        """
+        Generate random 3D rotation angles in degrees
+
+        :param n: The number of rotation angles to generate
+        :param seed: Random integer seed to use. If None,
+            the current random state is used.
+        :return: A n-by-3 ndarray of rotation angles in degrees
+        """
+        # Generate random rotation angles, in radians
+        angles = np.zeros((n, 3), dtype=self.dtype)
+        with Random(seed):
+            angles = np.column_stack(
+                (
+                    np.random.random(n) * 2 * np.pi,
+                    np.arccos(2 * np.random.random(n) - 1),
+                    np.random.random(n) * 2 * np.pi,
+                )
+            )
+        # Return random rotation angles in degrees
+        return np.rad2deg(angles).astype(self.dtype)
+
+    @property
+    def rot_matrices(self):
+        """
+        :return: Rotation matrices as a n x 3 x 3 array
+        """
+        return self._rotations.as_matrix().astype(self.dtype)
+
+    @rot_matrices.setter
+    def rot_matrices(self, values):
+        """
+        Set rotation matrices
+
+        :param values: Rotation matrices as a n x 3 x 3 array
+        :return: None
+        """
+        self._rotations = sp_rot.from_matrix(values.astype(self.dtype))
+        self.data = values.astype(self.dtype)
+
+    @property
+    def angles(self):
+        """
+        :return: Rotation angles in degrees, as a n x 3 array
+        """
+        return self._rotations.as_euler(self.rot_seq, degrees=True).astype(self.dtype)
+
+    @angles.setter
+    def angles(self, values):
+        """
+        Set rotation angles in degrees
+
+        :param values: Rotation angles in degrees, as a n x 3 array
+        :return: None
+        """
+        self._rotations = sp_rot.from_euler(
+            self.rot_seq, values.astype(self.dtype), degrees=True
+        )
+
+    @property
```

Comment:
Remove.

Reply:
Removed in commit 0c8f0b8.

Reply:
No still there.

Reply:
I removed all @property  except for .T which is used to represent transpose matrices instead of a function .T() in a way similar to numpy object or BlkDiagMatrix object. Remove too?

Reply:
Yes but transpose here is ambigious. Does it mean transpose the entire array to (3, 3, n) or does it mean to transpose each individual rotation matrix? The NumPy interpretation would imply the former. It's clearer to just have the method `invert` (you can't invert a (n, 3, 3) array).

Reply:
So yes. Remove.

Reply:
`T ` was removed in commit  0a4768d.